BUG: Fix for .str.replace with repl function

hzpc-joostk opened the pull request:
.str.replace now accepts a callable (function) as replacement string. It now raises a TypeError when repl is not string like nor a callable. Docstring updated accordingly.

  - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry
- [x] closes #15055

Commit:
BUG: Fix for .str.replace with repl function

.str.replace now accepts a callable (function) as replacement string. It now raises a TypeError when repl is not string like nor a callable. Docstring updated accordingly.

chris-b1 commented:
Looks good, can you add tests and a whatsnew entry?

jreback commented on pandas/core/strings.py:
```diff
@@ -303,8 +303,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
-    repl : string
-        Replacement sequence
+    repl : string or function
+        Replacement string or a function, which passed the match object and 
```
function -> callable

Commit:
Update .str.replace docstring

As suggested by @jreback

Commit:
add tests for .str.replace with callable repl

See #15055 and #15056

hzpc-joostk commented:
Thanks, @chris-b1 and @jreback. Where should I put the whatsnew entry? Can I just pick an empty spot in doc/source/whatsnew/v0.20.0.txt?

    - Bug in ``.str.replace`` now accepts a callable replacement which is passed to ``re.sub`` (:issue:`15055`)

This maybe a new feature or improvement rather than a bug, right?

chris-b1 commented:
Yeah, I think new features is fine - I had marked the issue a  bug because this would have worked in older versions, although basically undocumented.

Commit:
Bug fix in .str.replace type checking done wrong

Dull mistake

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -435,6 +435,12 @@ def test_replace(self):
+        # GH 15055, callable repl
+        repl = lambda m: m.group(0).swapcase()
+        result = values.str.replace('[a-z][A-Z]{2}', repl, n=2)
+        exp = Series([u('foObaD__baRbaD'), NA])
```
test where the callable has different than 1 arg (should be error)

test using named field in re (and callable works correctly)

hzpc-joostk replied:
Thanks, @jreback . Named field/group is a good idea.

If the callable raises a `TypeError` because of a wrong number of arguments, this is caught in [`pandas.core.strings._map`](https://github.com/hzpc-joostk/pandas/blob/pandas-GH15055-patch-1/pandas/core/strings.py#L175). 

We could overcome this by either inspecting the callable signature: `inspect.signature` or trying with a dummy Regex Match object. But this might have the side effect that the callable is called once before doing the replace on the array, confusing users with some kind of counter within the callable.

Sorry about the checks going wrong. It seems I am too careless with checking my commits. 😕

hzpc-joostk replied:
With `inspect` I could do

```py
n_empty_args = sum(1 for p in inspect.signature(repl).parameters.values() 
                   if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD, p.KEYWORD_ONLY) 
                   and p.default is p.empty)

# or better readable syntax

signature = inspect.signature(repl)
empty_args = [p for p in signature.parameters.values() 
              if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD, p.KEYWORD_ONLY) 
              and p.default is p.empty]
n_empty_args = len(empty_args)

# followed by

if n_empty_args > 1:
    raise TypeError("callable should have at most one positional without a default")
```

`n_empty_args != 1` might work, unless the first arg has a default for some reason.

jreback replied:
this is too complicated, instead *maybe* you can introspect the actual error message in ``_map`` and if its a certain class of errors, just re-raise the TypeError

Commit:
Fix testing bug for .str.replace

I copied a previous test as a draft, assumed that I could reuse `values` 
but did not mention it was redefined in the unicode and GH 13438 tests. 
Now corrected. Just redefine `values`. I could have put my test before 
the unicode tests but this keeps the order.

Commit:
Add whatsnew for .str.replace with callable repl

See #15056

Commit:
Reraise TypeError only with wrong number of args

This influences all StringMethods were a used passes a callable to a StringMethod, 
e.g. `.str.replace`, that expects a wrong number of arguments.

See #15056

Commit:
Add test for .str.replace with regex named groups

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -435,6 +435,31 @@ def test_replace(self):
                     values = klass(data)
                     self.assertRaises(TypeError, values.str.replace, 'a', repl)
+        
+        ## GH 15055
```
can you make a separate test

jreback commented on pandas/core/strings.py:
```diff
@@ -167,7 +167,14 @@ def _map(f, arr, na_mask=False, na_value=np.nan, dtype=object):
             result = lib.map_infer_mask(arr, f, mask.view(np.uint8), convert)
-        except (TypeError, AttributeError):
+        except (TypeError, AttributeError) as e:
+            re_missing = (r'missing \d+ required (positional|keyword-only) '
```
can you add a 1-line comment on what types of exceptions you are re-raising here (imagine you are a future reader and are like what?)

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -23,6 +23,7 @@ New features
 ~~~~~~~~~~~~
 
 - Integration with the ``feather-format``, including a new top-level ``pd.read_feather()`` and ``DataFrame.to_feather()`` method, see :ref:`here <io.feather>`.
+- ``.str.replace`` now accepts a callable replacement which is passed to ``re.sub`` (:issue:`15055`)
```
callable, as a replacement, which ....

jreback commented on pandas/core/strings.py:
```diff
@@ -318,9 +326,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
-        raise TypeError("repl must be a string")
-    use_re = not case or len(pat) > 1 or flags
+    if not (is_string_like(repl) or callable(repl)):
+        raise TypeError("repl must be a string or function")
```
function -> callable

jreback commented on pandas/core/strings.py:
```diff
@@ -303,8 +310,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
-        Replacement sequence
+    repl : string or callable
+        Replacement string or a callable, it's passed the match object and 
+        must return a replacement string to be used. See :func:`re.sub`.
```
add the version (0.20.0) after callable

jreback commented on pandas/core/strings.py:
```diff
@@ -303,8 +310,9 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
+        Replacement string or a callable, it's passed the match object and 
+        must return a replacement string to be used. See :func:`re.sub`.
     n : int, default -1 (all)
         Number of replacements to make from start
```
can you add some examples to the doc-string?

also would be nice to have an example in text.rst

jreback reviewed (changes requested):
lgtm. just some added documenations / minor comments. ping on green.

Commit:
improve test .str.replace with callable

- separate test function
- fix missing assertRaisesRegex(p)

Commit:
improve .str.replace with callable

- add inline comments
- add examples with string and callable in __doc__ and text.rst doc
- extend method summary Series.str.replace in text.rst doc

hzpc-joostk commented:
Thanks. I added changes based on your review. Only building with python=2.7 fails. I have little to no experience with Python 2x. I didn't caught the TypeError in Python 2:

jreback commented:
so

- some linting issues, use ``git diff master | flake8 --diff`` to show locally
- setup a venv or conda env with py2.7 to step thru the issues there.

Generally python 3 does the right thing with exceptions, while py2 may not be re-raising exceptions correctly (so your catching may not work here).

So you may need to have a slightly different strategy in py2 and py3 for catching the invalid args.

If you get stuck, ping.

hzpc-joostk commented:
Currently I'm designing new regexes for both py27 and py3k. I'll look into the linting.

jreback commented:
@hzpc-joostk 

>Currently I'm designing new regexes for both py27 and py3k. I'll look into the linting.

what do you mean? you can prob do something very straightforward, you don't even need regexes, you are just trying to find out if the error message if related to the calling of the function (and not the dtype of the data).

Commit:
fix bug catch TypeError with wrong number of args

The regex filtering TypeErrors with argument number problems is now unified and suitable for Python 2.7

Commit:
Update v0.20.0.txt

hzpc-joostk commented:
@jreback earlier you reviewed a part of my code:
> test where the callable has different than 1 arg (should be error)

In `.core.strings._map` these string methods are mapped. On both `TypeError` and `AttributeError`, `NaN` is returned instead. Using regex, implemented in commit 27065a29ba883f79499e4978872611f23d606f7e, only `TypeError`s regarding wrong number of args are reraised. However, these regexes only worked for Python 3. With e15dcdf544f02de04573b942132186d8ea397676, these regexes are unified into one sophisticated (complex) regex and work for Python 2 (tested).

tl;dr the tests should pass, as this now raises an error.

jreback commented:
@hzpc-joostk lgtm. ping on green.

hzpc-joostk commented:
Now fixing linting issues: 

    C:\Users\joostk\Documents\git\pandas>git diff 6ecc43d83801bd0337b771390ea5f4063689f275 | flake8 --diff
    .\pandas/core/strings.py:313:75: W291 trailing whitespace
    .\pandas/core/strings.py:314:70: W291 trailing whitespace
    .\pandas/core/strings.py:332:65: W291 trailing whitespace
    .\pandas/core/strings.py:341:69: W291 trailing whitespace
    .\pandas/core/strings.py:342:71: W291 trailing whitespace
    .\pandas/tests/test_strings.py:438:1: W293 blank line contains whitespace
    .\pandas/tests/test_strings.py:439:1: D102 Missing docstring in public method
    .\pandas/tests/test_strings.py:440:9: E266 too many leading '#' for block comment
    .\pandas/tests/test_strings.py:468:1: D102 Missing docstring in public method

Commit:
fix linting issues

- trailing whitespace
- whitespace empty line
- too many '#' in comment

hzpc-joostk commented:
@jreback Thanks for your input and help. All changes made. Waiting for CI... Besides that: all green!

jreback commented:
https://travis-ci.org/pandas-dev/pandas/jobs/194493716

@hzpc-joostk 

so you can use

```
if PY2:
    # match py2 excetion
else:
    # match PY3 exception
```

and as I said you can do a minimal matching, because sometimes these exception messages change in versions (e.g. 3.5 -> 3.6 for example, maybe not in this case, but the wording can change). You are just generically interested in a specific type of error message (and not its specific wording per se)

Commit:
simplify .str.replace TypeError reraising and test

jreback commented:
thanks @hzpc-joostk 

nicely done!

hzpc-joostk commented:
Thanks @jreback 

Awesome. My first (accepted) PR on GitHub. 🍾 Cheers
